src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3152,6 +3153,312 @@ void MacroAssembler::check_klass_subtype(Register sub_klass,
+
+  bind(L_exit);
+  BLOCK_COMMENT("} repne_scan");
+}
+
+// Ensure that the inline code and the stub are using the same registers.
+#define LOOKUP_SECONDARY_SUPERS_TABLE_REGISTERS                 \
+do {                                                            \
+  assert(r_super_klass  == Z_ARG1                            && \
+         r_array_base   == Z_ARG5                            && \
+         r_array_length == Z_ARG4                            && \
+        (r_array_index  == Z_ARG3 || r_array_index == noreg) && \
+        (r_sub_klass    == Z_ARG2 || r_sub_klass   == noreg) && \
+        (r_bitmap       == Z_R10  || r_bitmap      == noreg) && \
+        (r_result       == Z_R11  || r_result      == noreg), "registers must match s390.ad"); \
+} while(0)
+
+// Note: this method also kills Z_R1_scratch register; Beware of Register clobbering in future
+void MacroAssembler::lookup_secondary_supers_table(Register r_sub_klass,
+                                                   Register r_super_klass,
+                                                   Register r_temp1,
+                                                   Register r_temp2,
+                                                   Register r_temp3,
+                                                   Register r_temp4,
+                                                   Register r_result,
+                                                   u1 super_klass_slot) {
+  NearLabel L_done, L_failure;
+
+  BLOCK_COMMENT("lookup_secondary_supers_table {");
+
+  const Register
+    r_array_base   = r_temp1,
+    r_array_length = r_temp2,
+    r_array_index  = r_temp3,
+    r_bitmap       = r_temp4;
+
+  LOOKUP_SECONDARY_SUPERS_TABLE_REGISTERS;
+
+  z_lg(r_bitmap, Address(r_sub_klass, Klass::bitmap_offset()));
+
+  // First check the bitmap to see if super_klass might be present. If
+  // the bit is zero, we are certain that super_klass is not one of
+  // the secondary supers.
+  u1 bit = super_klass_slot;
+  int shift_count = Klass::SECONDARY_SUPERS_TABLE_MASK - bit;
+
+  z_sllg(r_array_index, r_bitmap, shift_count); // take the bit to 63rd location
+
+  clear_reg(r_result, true /* whole_reg */, false /* set_cc */); // r_result = 0, let's hope that search will be a success
+
+  // We test the MSB of r_array_index, i.e., its sign bit
+  testbit(r_array_index, 63);
+  z_bfalse(L_failure); // if not set, then jump!!!
+
+  // We will consult the secondary-super array.
+  z_lg(r_array_base, Address(r_sub_klass, Klass::secondary_supers_offset()));
+
+  // The value i in r_array_index is >= 1, so even though r_array_base
+  // points to the length, we don't need to adjust it to point to the
+  // data.
+  assert(Array<Klass*>::base_offset_in_bytes() == wordSize, "Adjust this code");
+
+  // Get the first array index that can contain super_klass.
+  if (bit != 0) {
+    pop_count_long(r_array_index, r_array_index, Z_R1_scratch); // all the registers are hardcoded so should be fine
```

Comment:
This comment is also rather baffling. You seem to be concerned about something, but what? `pop_count_long` doesn't cause any particular risk, does it?

Reply:
For machines older than `Z15`, `pop_count_long` clobbers `Z_R1_scratch` register.  That's why I added it there.

Reply:
Better then just to say what matters: "NB: May clobber Z_R1_scratch" or "Clobbers Z_R1_scratch on older machines."
"Should be fine" is just confusing because "should" is tentative, where you need to be certain in comments.